### PR TITLE
Add the ability to construct a callback function

### PR DESCRIPTION
**Pull request by domenic**
This helps move some of the heavy lifting into IDL, to avoid issues like https://github.com/whatwg/html/issues/2381.

***
[Preview](https://s3.amazonaws.com/pr-preview/heycam/webidl/construct-helper.html) | [Diff](https://s3.amazonaws.com/pr-preview/heycam/webidl/fe42994...d445e3f.html)

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -12374,6 +12374,56 @@ described in the previous section).
...
+<div algorithm>
+
+    To <dfn id="construct-a-callback-function" export>construct</dfn> a
+    [=callback function type=] value |callable| with
+    a list of arguments |arg|<sub>0..|n|−1</sub>, each of which is either
+    an IDL value or the special value “missing” (representing a missing optional argument),
+    perform the following steps.
+    These steps will either return an IDL value or throw an exception.
+
+    1.  Let |completion| be an uninitialized variable.
```
Comment (bzbarsky) [inline index.bs]: So this algorithm is pretty similar to <https://heycam.github.io/webidl/#invoke-a-callback-function>.  I wonder whether it's worth trying to factor parts of it out...  I guess the main similarity is the preparation (after the basic sanity-check, which is different) and the argument conversions, right?

Maybe that's not enough to worry about it.

File: index.bs [hunk 2]
```diff
@@ -12374,6 +12374,56 @@ described in the previous section).
...
+    1.  If |callResult| is an abrupt completion, set |completion| to
+        |callResult| and jump to the step labeled <a href="#construct-return"><i>return</i></a>.
+    1.  Set |completion| to the result of [=converted to an IDL value|converting=]
+        |callResult|.\[[Value]] to an IDL value of the same type as the operation’s
+        return type.
+    1.  <i id="construct-return">Return:</i> at this
+        point |completion| will be set to an ECMAScript completion value.
+        1.  [=Clean up after running a callback=] with |stored settings|.
+        1.  [=Clean up after running script=] with |relevant settings|.
+        1.  Return |completion|.
```
Comment (bzbarsky) [inline index.bs]: Up front we said "These steps will either return an IDL value or throw an exception."  But we're actually returning an ES completion value... I'm not sure how to reconcile those.

Of course I'm also not sure in what sense "converting callResult.[[Value]] to an IDL value" returns an ES completion value...

File: index.bs (hunk 1 above)
Comment (domenic) [inline index.bs]: I factored out the arguments conversion, which I think factors the most cleanly. The rest varies a bit too much IMO.

File: index.bs (hunk 2 above)
Comment (domenic) [inline index.bs]: This is a preexisting "problem" that ideally I'd like to not fix in this PR.

Personally I think it is OK for specs to treat returning completion values and traditional return/throw flow the same way, without an explicit conversion step. (And similarly, treating return/flow throw as returning completion values.) It's a bit strange that the ES completion record holds a Web IDL value, but IMO fine.

Comment (domenic) [inline index.bs]: We could mention this equivalence and our expanded use of ES completion values in the conventions section.

File: index.bs [hunk 3]
```diff
@@ -12166,14 +12166,37 @@ a [=callback interface=] that:
...
+    1.  While |i| &lt; |args|'s [=list/size=]:
+        1.  If |args|[|i|] is the special value “missing”, then [=list/append=]
+            <emu-val>undefined</emu-val> to |esArgs|.
+        1.  Otherwise, |args|[|i|] is an IDL value:
+            1.  Let |convertResult| be the result of [=converted to an ECMAScript value|converting=]
+                |args|[|i|] to an ECMAScript value. Rethrow any exceptions.
+            1.  [=list/Append=] |convertResult| to |esArgs|.
+            1.  Set |count| to |i| + 1.
+        1.  Set |i| to |i| + 1.
+    1.  [=list/Remove=] all items from |esArgs| except the first |count|, so that |esArgs|'s
```
Comment (tobie) [inline index.bs]: "Truncate" is a lot clearer to read, even though it's not defined in infra. Should it be defined there? (see https://github.com/whatwg/infra/issues/100)

File: index.bs [hunk 4]
```diff
@@ -12166,14 +12166,37 @@ a [=callback interface=] that:
...
+    1.  Let |count| be 0.
+    1.  While |i| &lt; |args|'s [=list/size=]:
+        1.  If |args|[|i|] is the special value “missing”, then [=list/append=]
+            <emu-val>undefined</emu-val> to |esArgs|.
+        1.  Otherwise, |args|[|i|] is an IDL value:
+            1.  Let |convertResult| be the result of [=converted to an ECMAScript value|converting=]
+                |args|[|i|] to an ECMAScript value. Rethrow any exceptions.
+            1.  [=list/Append=] |convertResult| to |esArgs|.
+            1.  Set |count| to |i| + 1.
+        1.  Set |i| to |i| + 1.
```
Comment (tobie) [inline index.bs]: Increment? (see whatwg/infra#99).

File: index.bs [hunk 5]
```diff
@@ -12374,6 +12373,42 @@ described in the previous section).
...
+    1.  Let |stored settings| be |callable|'s [=callback context=].
+    1.  [=Prepare to run script=] with |relevant settings|.
+    1.  [=Prepare to run a callback=] with |stored settings|.
+    1.  Let |esArgs| be the result of [=Web IDL arguments list/converting=] |args| to an ECMAScript
+        arguments list. If this throws an exception, set |completion| to the completion value
+        representing the thrown exception and jump to the step labeled
+        <a href="#construct-return"><i>return</i></a>.
+    1.  Let |callResult| be [=Construct=](|F|, |esArgs|).
+    1.  If |callResult| is an abrupt completion, set |completion| to
+        |callResult| and jump to the step labeled <a href="#construct-return"><i>return</i></a>.
```
Comment (tobie) [inline index.bs]: I know we use steps labelled "return" all over the place. I still find that icky. Of course we shouldn't change it here.

File: index.bs (hunk 2 above)
Comment (tobie) [inline index.bs]: Yeah, we seem to mix that up all over the place. I guess we should define it more clearly. Maybe that's yet another thing for the infra standard?

Comment (bzbarsky) [inline index.bs]: I think having an explicit equivalence here is fine, fwiw.  We should just do that.

**Pull request merged**